deep_qa/training/optimizers.py

```diff
@@ -15,18 +15,114 @@
 of things really well. It just has a few quirks...
 """
 from typing import Union
+
+from keras import backend as K
+
 from keras.optimizers import SGD, RMSprop, Adagrad, Adadelta, Adam, Adamax, Nadam
 from ..common.params import Params
 
-optimizers = {  # pylint: disable=invalid-name
-        'sgd': SGD,
-        'rmsprop': RMSprop,
-        'adagrad': Adagrad,
-        'adadelta': Adadelta,
-        'adam': Adam,
+
+if K.backend() == "tensorflow":
+    from keras.optimizers import Optimizer
+    from tensorflow.python.training import adagrad, adadelta, adam, gradient_descent, rmsprop, optimizer
+    from tensorflow import clip_by_value, clip_by_norm
+
+
+    class TFOptimizer(Optimizer):
+        """
+        Wrapper class for native TensorFlow optimizers. This exists already in Keras,
+        but it doesn't support the same API as the other Keras optimisers, which this
+        class fixes by allowing gradient clipping by value or by norm.
+        """
+        def __init__(self, optimizer, **kwargs):
+
+            self.clip_norm = kwargs.pop("clipnorm", None)
+            self.clip_value = kwargs.pop("clipvalue", None)
+
+            if kwargs != {}:
+                raise TypeError('Unexpected keyword argument/s '
+                                'passed to optimizer: ' + str(kwargs))
+
+            self.optimizer = optimizer
+            self.iterations = K.variable(0., name='iterations')
+            self.updates = []
+
+        def get_updates(self, params, constraints, loss):
+            if constraints:
+                raise ValueError('TF optimizers do not support '
+                                 'weights constraints. Either remove '
+                                 'all weights constraints in your model, '
+                                 'or use a Keras optimizer.')
+            grads = self.optimizer.compute_gradients(loss, params)
+
+            if self.clip_norm is not None:
+                grads = [clip_by_norm(grad, self.clip_norm) for grad in grads]
```

Comment:
I'm not sure how keras implements this... but it's generally recommended to use tf.clip_by_global_norm (see https://arxiv.org/abs/1211.5063)